File: cpp/graphAlgorithms/ecolor/ecolor_g.cpp
/** @file ecolor_g.cpp
 * 
 *  @author Jon Turner
 *  @date 2015
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "ecolor_g.h"
#include "mdmatch_f.h"
#include "mdmatch.h"

namespace grafalgo {

/** Find a minimum edge coloring in a bipartite graph using Gabow's algorithm.
 *  @param g is a reference to the graph
 *  @param color is an array indexed by an edge number; on return
 *  color[e] is the color assigned to edge e
 *  @return the number of colors used or 0 if the graph is not bipartite
 */
ecolor_g::ecolor_g(Graph& g, int color[]) {
	// initialize data structures
	gp = &g;
	cg = new Graph(g.n(),g.M());
	cg->copyFrom(g);
	this->color = color;
	match = new List_g<edge>(g.n()/2);
	euler = new Djsets_cl(g.M());
	handle = new List_g<edge>(g.M());
	start = new List_d(g.n());

	nextColor = 1;
	rColor(g.maxDegree());
}

ecolor_g::~ecolor_g() {
	delete match; delete euler; delete handle; delete start;
}

/** Recursive helper function.
 *  Colors the current graph cg, modifying it as necessary, along the way.
 *  Note: cg uses same vertex numbers and edge numbers as original graph.
 *  Also uses global data structures euler and handle, within eulerPartition
 *  method.
 *  @param Delta is the max vertex degree in the current graph cg.
 */
void ecolor_g::rColor(int Delta) {
	// if cg is a matching, just color its edges
	if (Delta == 1) {
		for (edge e = cg->first(); e != 0; e = cg->next(e))
			color[e] = nextColor;
		nextColor++;
		return;
	}

	if ((Delta&1) == 1) {
		// find matching in cg that includes every max degree vertex
		match->clear();
		mdmatch_f(*cg, *match);

		// now, color edges in matching and remove from cg
		for (index x = match->first(); x != 0; x = match->next(x)) {
			edge e = match->value(x);
			color[e] = nextColor; cg->remove(e);
		}
		nextColor++;

		Delta--; // decrement max degree
	}
	// cg now has even maximum degree
	int m = cg->m();

	// find Euler partition and return in euler and handle;
	// on completion, cg has no edges
	eulerPartition();

	// rebuild cg using half of the edges, place others in list
	List_g<edge> L(m/2); 
	while (!handle->empty()) {
		edge e = handle->value(handle->first()); handle->removeFirst();
		edge ee = e; bool odd = true;
		do {
			if (odd) cg->joinWith(gp->left(ee), gp->right(ee), ee);
			else L.addLast(ee);
			odd = not odd;
			edge tmp = ee;
			ee = euler->next(ee);
			if (tmp != e) euler->remove(tmp);
		} while (ee != e);
	}
	// note: euler and handle now ready for re-use

	// recursive calls on the two subgraphs
	rColor(Delta/2);
	cg->clear();
	for (index x = L.first(); x != 0; x = L.next(x)) {
		edge e = L.value(x);
		cg->joinWith(gp->left(e), gp->right(e), e);
	}
	rColor(Delta/2);
}

/** Find an Euler partition in the current graph.
 *  The partition is returned in the euler/handle data structures.
 *  Specifically, handle contains the "first" edge of some edge set
 *  in the partition. The edge sets are defined by circular lists in
 *  the euler data structure.
 */
void ecolor_g::eulerPartition() {
	// make list of start vertices, beginning with those of odd degree
	for (vertex u = 1; u <= cg->n(); u++) {
		if ((cg->degree(u)&1) == 1) start->addFirst(u);
		else if (cg->firstAt(u) != 0) start->addLast(u);
	}

	// traverse paths/cycles from start vertices
	// place edges in separate lists
	while (!start->empty()) {
		vertex s = start->first(); start->removeFirst();
		edge e = cg->firstAt(s);
		if (e == 0) continue;
		vertex v = s; edge ee = e;
		do {
			if (ee != e) euler->join(ee,e);
			v = cg->mate(v,ee);
			cg->remove(ee);
			ee = cg->firstAt(v);
		} while (ee != 0);
		handle->addLast(e);
		if (cg->firstAt(s) != 0) start->addLast(s);
	}
	// note: start now empty, ready for re-use
}

} // ends namespace

File: js/graphAlgorithms/ecolor/ecolorG.mjs
/** @file ecolorG.mjs
 *
 *  @author Jon Turner
 *  @date 2022
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import { assert, EnableAssert as ea } from '../../common/Assert.mjs';
import List from '../../dataStructures/basic/List.mjs';
import ListPair from '../../dataStructures/basic/ListPair.mjs';
import ListSet from '../../dataStructures/basic/ListSet.mjs';
import Graph from '../../dataStructures/graphs/Graph.mjs';
import findSplit from '../misc/findSplit.mjs';
import mdmatchG from '../vmatch/mdmatchG.mjs';

let g;			// shared reference to input graph
let io;         // ListPair defining bipartition on g
let wg;			// current working subgraph

let degree;		// degree[u] is the degree of u in wg
let active;		// List of vertices with non-zero degree wg
let subsets;	// vertex subsets that define bipartition
let emap;		// temp array used to map edge numbers in rcolor

let nextColor;  // next color to apply to edges
let color;		/// color[e] is color of edge e

let trace;
let traceString;

let steps;		// number of steps (inner loops)
let matches;	// number of matchings found


/** Compute a coloring of a bipartite graph using Gabow's algorithm.
 *  @param G is an undirected bipartite graph to be colored
 *  @param trace causes a trace string to be returned when true
 *  @return a pair [color, ts, stats] where color is an array of
 *  edge colors, ts is a trace string and stats is a statistics object
 */
export default function ecolorG(G, traceFlag=false) {
	// initialize data structures
	g = G; trace = traceFlag; trace = 1;
	ea && assert(g.hasBipartition);

	wg = new Graph(g.n, g.edgeRange); wg.setBipartition(g.getBipartition());
	degree = new Int32Array(g.n+1);
	active = new List(g.n);
	emap = new Int32Array(g.m+1);
	color = new Int32Array(g.edgeRange+1);

	for (let e = g.first(); e; e = g.next(e)) addEdge(e);

	traceString = '';
	if (trace) traceString += 'euler partitions and color sets\n'
	steps = matches = 0;

	nextColor = 1;
	rcolor(g.maxDegree());

	if (trace)
		traceString += g.toString(1,(e,u)=>`${g.x2s(g.mate(u,e))}:${color[e]}`);
	return [color, traceString, {'matches': matches, 'steps': steps }];
}

/** Recursive helper function.
 *  Colors the working graph wg, modifying it as necessary, along the way.
 *  Note: wg uses same vertex numbers and edge numbers as original graph.
 *  Also uses global data structures paths and starters, within eulerPartition
 *  method.
 *  @param Delta is the max vertex degree in the working graph wg.
 */
function rcolor(Delta) {
	// if wg is a matching, color its edges in g and remove them from wg
	if (Delta == 1) {
		if (trace) traceString += nextColor + ':';
		for (let e = wg.first(); e; e = wg.first()) {
			color[e] = nextColor; dropEdge(e);
			if (trace) traceString += ' ' + g.e2s(e,0,1);
			steps++;
		}
		if (trace) traceString += '\n';
		nextColor++;
		return;
	}

	// for odd Delta, extract matching on all degree Delta vertices,
	// color its edges and remove them from wg
	if (Delta & 1) {
		// first build compact version of wg with no isolated vertices
		let cwg = new Graph(active.length,g.m);
		let io = new ListPair(active.length);
		let u = 1;
		for (let v = active.first(); v; v = active.next(v)) {
			active.value(v,u); // use value to map wg's vertices to cwg's
			if (wg.isInput(v)) io.swap(u);
			u++; steps++;
		}
		cwg.setBipartition(io);
		for (let e = wg.first(); e; e = wg.next(e)) {
			let [u,v] = [wg.input(e),wg.output(e)];
			let [uu,vv] = [active.value(u), active.value(v)];
			let ee = cwg.join(uu,vv); emap[ee] = e;
			steps++;
		}

		// get max degree matching of cwg
		let [match,,stats] = mdmatchG(cwg);
		matches++; steps += stats.steps;
		if (trace) traceString += nextColor + ':';
		for (let e = match.first(); e; e = match.next(e)) {
			let ee = emap[e]; color[ee] = nextColor; dropEdge(ee);
			if (trace) traceString += ` ${g.e2s(ee,0,1)}`;
			steps++;
		}
		if (trace) traceString += '\n';
		nextColor++; Delta--;
	}

	// wg now has even maximum degree
	let [part1,part2] = eulerPartition();
	if (trace) {
		traceString += `${Delta/2} [`;
		for (let i = 0; i < part1.length; i++)
			traceString += `${i>0 ? ' ' : ''}${g.e2s(part1[i],0,1)}`;
		traceString += ']\n  [';
		for (let i = 0; i < part2.length; i++)
			traceString += `${i>0 ? ' ' : ''}${g.e2s(part2[i],0,1)}`;
		traceString += ']\n';
	}
	for (let i = 0; i < part1.length; i++) addEdge(part1[i]);
	rcolor(Delta/2);
	for (let i = 0; i < part2.length; i++) addEdge(part2[i]);
	rcolor(Delta/2);
	// wg is now empty
}

/** Add an edge to working graph.
 *  Update vertex degrees of endpoints and active vertex list.
 *  @param e is an edge in g to be added to wg
 */
function addEdge(e) {
	let [u,v] = [g.left(e),g.right(e)];
	wg.join(u,v,e);
	if (!degree[u]) active.enq(u);
	if (!degree[v]) active.enq(v);
	degree[u]++; degree[v]++;
}

/** Drop an edge from working graph.
 *  Update vertex degrees of endpoints and active vertex list.
 *  @param e is an edge in g to be removed from wg
 */
function dropEdge(e) {
	wg.delete(e);
	let [u,v] = [g.left(e),g.right(e)];
	degree[u]--; degree[v]--;
	if (!degree[u]) active.delete(u);
	if (!degree[v]) active.delete(v);
}
	
/** Find an Euler partition in the working graph.
 *  The partition is returned in Lists part1 and part2.
 */
function eulerPartition() {
	// first bring odd degree vertices to front of active list
	let nextu; let odds = 0;
	for (let u = active.first(); u; u = nextu) {
		nextu = active.next(u);
		if (degree[u]&1) {
			active.delete(u); active.push(u); odds++;
		}
		steps++;
	}

	let part1 = new Int32Array(wg.m&1 ? (wg.m+1)/2 : wg.m/2);
	let part2 = new Int32Array(wg.m&1 ? (wg.m-1)/2 : wg.m/2);
	let n1 = 0; let n2 = 0;
	while (!active.empty()) {
		let first = active.first();
		let v = first; let e = wg.firstAt(v);
		let balance = (n1 <= n2 ? 1 : 0);
		do {
			if (balance) part1[n1++] = e;
			else         part2[n2++] = e;
			v = g.mate(v,e); dropEdge(e);
			e = wg.firstAt(v);
			balance = !balance; steps++;
		} while (e);
		if (active.contains(first)) {
			// first no longer has odd degree, so move to end of active
			active.delete(first); active.enq(first);
		} 
	}
	return [part1,part2];
}
